python/overlay_widget/shotgun_spinning_widget.py

```diff
@@ -0,0 +1,223 @@
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+
+import math
+
+from tank.platform.qt import QtCore, QtGui
+
+# load resources
+from .ui import resources_rc # noqa
+
+
+class ShotgunSpinningWidget(QtGui.QWidget):
+    """
+    Overlay widget that can be placed on top over any QT widget.
+    Once you have placed the overlay widget, you can use it to
+    display a spinner or report progress in the form of an arc that goes
+    from 0 to 360 degrees.
+    """
+
+    MODE_OFF = 0
+    MODE_SPIN = 1
+    MODE_PROGRESS = 2
+
+    # Indicates how many times per second does the spinner update. This means every 40ms.
+    _UPDATES_PER_SECOND = 25
+
+    def __init__(self, parent):
+        """
+        :param parent: Widget to attach the overlay to
+        :type parent: :class:`PySide.QtGui.QWidget`
+        """
+        QtGui.QWidget.__init__(self, parent)
+
+        # turn off the widget
+        self.setVisible(False)
+        self._mode = self.MODE_OFF
+
+        # setup spinner timer
+        self._timer = QtCore.QTimer(self)
+        self._timer.timeout.connect(self._on_animation)
+
+        # This is the current spin angle
+        self._spin_angle = 0
+        # This is where we need to scroll to.
+        self._spin_angle_to = 0
+        # This is where we were told last time where we need to spin to
+        self._previous_spin_angle_to = 0
+        # This counts how many times we've ticked in the last second to know how big the heartbeat
+        # needs to be.
+        self._heartbeat = 0
+
+        self._sg_icon = QtGui.QPixmap(":/tk_framework_qtwidgets.overlay_widget/sg_logo.png")
+
+    ############################################################################################
+    # public interface
+    def start_spin(self):
+        """
+        Enables the overlay and shows an animated spinner.
+
+        If you want to stop the spinning, call :meth:`hide`.
+        """
+        self._timer.start(40)
+        self.setVisible(True)
+        self._mode = self.MODE_SPIN
+
+    def start_progress(self):
+        """
+        Enables the overlay and shows an animated progress arc.
+
+        If you want to stop the progress, call :meth:`hide`.
+        """
+        self.setVisible(True)
+        self._timer.start(1000 / self._UPDATES_PER_SECOND)
+        self._mode = self.MODE_PROGRESS
+        self._spin_angle_to = 0
+        self._spin_angle = 0
+
+    def report_progress(self, current):
+        """
+        Updates the widget current progress value.
+
+        :param current: New value for the progress arc. Must be between 0 (nothing) and 1 (complete).
```

Comment:
nitpick: maybe say between 0.0 and 1.0 instead to highlight the fact that they are floats.